### PR TITLE
Konsti rnd exploration

### PR DESCRIPTION
**Pull request by KonstiNik**
Implement Random Network Distillation (RND) as an intrinsic reward module. 

#### Summary of additions and changes

- The training and network building of RND is based on ZnNL

- Created a new module for intrinsic rewards
- Implement Random Network Distillation as a child class of the intrinsic reward module. 
- Implement two default  Configurations for RND
- Added Integration and Unit Test for the new module

#### Todo

- [x] All Tests Pass with local ZnNL implementation
- [x] Import ZnNL into SwarmRL @SamTov 
- [x] Write example API script (For the paper)
- [ ] Add TrajectoryInformation to utils to avoid cyclic imports

### CODE REVIEW

**Review by SamTov — COMMENTED** [review]
The PR seems really cool! I am looking forward to using it. Can you answer the question so I have a better idea of the flow. Otherwise, it looks great and well tested. We should probably also add a test inside of an RL run.

File: CI/espresso_tests/integration_tests/test_rnd_intrinsic_reward_deployment.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        x = nn.Dense(features=32)(x)
+        return x
+
+
+@dataclass
+class EpisodeDataDummy:
+    features: np.ndarray
+
+
+class RNDTest(ut.TestCase):
```
[the remark this answers is not in the document]
Comment (KonstiNik) [reply CI/espresso_tests/integration_tests/test_rnd_intrinsic_reward_deployment.py:42]: I am using the assert methods but I can also do that with pytest. Is there an advantage?

File: CI/espresso_tests/integration_tests/test_rnd_intrinsic_reward_deployment.py [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    def test_reward_decrease(self):
+        """
+        Test whether the intrinsic reward decreases over training.
+        """
+        # Define the RND model using simple training
+        config = RNDConfig(agent_network=self.network)
+        intrinsic_reward = RNDReward(config)
+
+        reward_before = intrinsic_reward.compute_reward(self.episode_data)
+        intrinsic_reward.update(self.episode_data)
```
[the remark this answers is not in the document]
Comment (KonstiNik) [reply CI/espresso_tests/integration_tests/test_rnd_intrinsic_reward_deployment.py:84]: Thanks

File: swarmrl/agents/actor_critic.py [hunk 3]
```diff
@@ -179,12 +189,20 @@ def calc_action(self, colloids: typing.List[Colloid]) -> typing.List[Action]:
         )
         chosen_actions = np.take(list(self.actions.values()), action_indices, axis=-1)
 
+        # Compute extrinsic rewards.
+        rewards = self.task(colloids)
```
[the remark this answers is not in the document]
Comment (KonstiNik) [reply swarmrl/agents/actor_critic.py:193]: Before was computed directly in the append function (below in the code). 
Now I am adding the intrinsic reward to it. So it needs to be computed.

File: swarmrl/intrinsic_reward/rnd_configs.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+from flax import linen as nn
+from znnl.distance_metrics import DistanceMetric, OrderNDifference
+from znnl.loss_functions import MeanPowerLoss
+from znnl.training_strategies.loss_aware_reservoir import LossAwareReservoir
+from znnl.training_strategies.simple_training import SimpleTraining
+
+from swarmrl.networks.network import Network
+
+
+class RNDArchitecture(nn.Module):
```
[the remark this answers is not in the document]
Comment (KonstiNik) [reply swarmrl/intrinsic_reward/rnd_configs.py:22]: This can be changed of course. 
The RND architecture can be passed to the config for target and predictor model. 
Per default it is just set to the one implemented there. Also to provide one example for a user.

File: swarmrl/intrinsic_reward/rnd_configs.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    def __post_init__(self):
+        """
+        Constructor initializing the training strategy.
+        """
+        self.training_strategy = LossAwareReservoir(
+            model=None,
+            loss_fn=MeanPowerLoss(order=2),
+            reservoir_size=self.reservoir_size,
+            latest_points=self.episode_length,
+        )
```
[the remark this answers is not in the document]
Comment (KonstiNik) [reply swarmrl/intrinsic_reward/rnd_configs.py:138]: It seems like black does not care. But i will add one.

File: swarmrl/intrinsic_reward/rnd_configs.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (KonstiNik) [reply swarmrl/intrinsic_reward/rnd_configs.py:20]: added info in [83a10ce](https://github.com/SwarmRL/SwarmRL/pull/75/commits/83a10cefd2c3f2caf48f4d7c7535407716bd75c9)

File: swarmrl/intrinsic_reward/rnd_configs.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (KonstiNik) [reply swarmrl/intrinsic_reward/rnd_configs.py:146]: added line  in [83a10ce](https://github.com/SwarmRL/SwarmRL/pull/75/commits/83a10cefd2c3f2caf48f4d7c7535407716bd75c9)

**Comment by KonstiNik** [comment]
So far, I have put the integration test for RND to the espresso integration tests. 
Should I put it into the unit test suite for SwarmRL? 
Or how do you handle integration tests for neural networks?

**Comment by SamTov** [comment]
> So far, I have put the integration test for RND to the espresso integration tests. Should I put it into the unit test suite for SwarmRL? Or how do you handle integration tests for neural networks? @SamTov

Good question, it depends on what you want to test. If you are using the espresso engine, then the test must go into the espresso integration tests, otherwise, it should just go into the normal test suite. The difference is that for espresso tests we need to call a different kernel. I try to keep all things out of espresso tests and just put a bigger test in there if necessary.

**Comment by SamTov** [comment]
I have added the extra dependency to the `setup.py`. If you want to install with ZnNL, run:

```sh
pip install swarmrl[rnd]
```
or, locally
```sh
pip install -e ".[rnd]"
```

I also looked at the install issues. In general, I am concerned about the set versions of neuraltangents in ZnNL which could cause some weird issues. Let's remove that and maybe also ipython but just because we also have Jupyter in there so there is some redundancy.

**Comment by KonstiNik** [comment]
The espresso test fails in 3 instances:
- The rod rotation, which is very likely due to an update of espresso. 
- The other two are wrong times in the time measurement. We get e.g. 100.8 s instead of 0.8 s. Does anyone of you @SamTov @christophlohrmann have an idea where this is coming from? In both cases it seems to take 100 s more than expected.

**Comment by SamTov** [comment]
Test issues should be resolved now. What is the resolution on your `TrajectoryInformation` comment in the initial PR description?

**Comment by SamTov** [comment]
Just a final comment to the PR. Using the `pip install "swarmrl[rnd]"` or 'pip install ".[rnd]"` results in issues satisfying the TF package version. This lands eventually on 2.9. However, this is arising, as far as I can see, due to conflicts with the NT package and probably something we require in SwarmRL. Is there a fixed version somewhere on the ZnNL side pinning to an old version?

File: requirements.txt [hunk 6]
```diff
@@ -1,3 +1,5 @@
+jax==0.4.23
```
Comment (SamTov) [inline requirements.txt:1]: I am not really in favour of a hard pinning of Jax libraries, considering how difficult they are for everyone to match depending on GPU and other issues. From my messing around in a clean conda environment, this was not necessary. Is there any reason you really need to do this? I am on board with a `>=` restriction to bar low versions but I am not a fan of this approach.

File: swarmrl/intrinsic_reward/random_network_distillation.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+            flax_module=RNDArchitecture(),
+            optimizer=rnd_config.optimizer,
+            input_shape=(1, *rnd_config.input_shape),
+        )
+        self.training_strategy.set_model(self.predictor_network)
+
+    @staticmethod
+    def _reshape_data(x: np.ndarray) -> np.ndarray:
+        """
+        Reshape the data to treat the n_steps and num_colloids equally.
```
Comment (SamTov) [inline swarmrl/intrinsic_reward/random_network_distillation.py:60]: Perhaps this comment can read "Flatten the first two axes of the array". I don't find the current wording clear. If I didn't read the code, I would not understand what "treat them equally" means.

**Review by SamTov — CHANGES_REQUESTED** [review]
I just have a concern over the Jax version pinning and then one minor point about a doc string. If the version issue is resolved, the PR can be merged.

File: swarmrl/intrinsic_reward/random_network_distillation.py (hunk 7 above)
Comment (KonstiNik) [reply swarmrl/intrinsic_reward/random_network_distillation.py]: changed in [981cd2e](https://github.com/SwarmRL/SwarmRL/pull/75/commits/981cd2e708a7bcc20657e7e1a0eb5453f6286d88)

File: requirements.txt (hunk 6 above)
Comment (SamTov) [reply requirements.txt:1]: Do you think we can pin it with ">=" or do we need an upper bound as well?

**Review by SamTov — APPROVED** [review]
LGTM
